text/000-rocq-findlib.md

```diff
@@ -0,0 +1,267 @@
+
+## Requiring a package
+
+When a package `my_package` is required via a build system, the metadata can
+be used to compute a set of transitive dependencies for it. Using `ocamlfind`,
+this can be done as follows.
+```
+ocamlfind query p -r -format "%p"
+```
+
+The obtained packages can then be sorted into two lists: one containing OCaml
+libraries (including Rocq plugins), and one containing Rocq theories. One may
+run the following command to determine whether a package is a theory or not.
+```
+ocamlfind query some_dep -format "%(rocqpath)"
+```
+The result is empty if `some_dep` is a standard OCaml package (possibly a Rocq
+plugin), and it gives a directory path for Rocq theories.
+
+The mapping for each theory package `some_theory` can then be computed using
+the following command.
+```
+ocamlfind query some_theory -format "-Q %d/rocq.d %(rocqpath)"
+```
+
+## Example: stdpp
+
+The packages for [stdpp](https://gitlab.mpi-sws.org/iris/stdpp) could end up
+being installed under `<PREFIX>/lib/rocq-stdpp/`, which would contains the
+following `META` file.
+```
+version = "dev"
+description = "An extended \"Standard Library\" for Rocq."
+rocqpath = "stdpp"
+requires = "rocq-stdlib"
+
+package "bitvector" (
+  directory = "bitvector"
+  version = "dev"
+  description = "A library for bitvectors based on std++."
+  rocqpath = "stdpp.bitvector"
+  requires = "rocq-stdpp"
+)
+
+package "unstable" (
+  directory = "unstable"
+  version = "dev"
+  description = "Unfinished std++ libraries."
+  rocqpath = "stdpp.unstable"
+  requires = "rocq-stdpp"
+)
+```
+
+Differences with the current `stdpp` setup:
+- Rocq-ified package names.
+- Use of sub-packages form `stdpp.bitvector` and `stdpp.unstable`, instead of
+  different Opam packages.
+
+## Example: coq-elpi
+
+```
+rocqpath = "elpi"
+requires = "rocq-stdlib coq-elpi.plugin"
+...
+
+package "plugin" (
+  directory = "plugin"
+  requires = "rocq-core.plugins.ltac rocq-core.vernac elpi"
+  ...
+)
+
+package "apps" (
+  directory "apps"
+
+  package "derive" (
+    directory "derive"
+    rocqpath = "elpi.apps.derive"
+    requires = "rocq-elpi rocq-elpi.apps.derive.elpi"
+    ...
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.derive.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  package "tc" (
+    directory "tc"
+    rocqpath = "elpi.apps.tc"
+    requires = "rocq-elpi rocq-elpi.apps.tc.elpi rocq-elpi.apps.tc.plugin"
+    ...
+
+    package "plugin" (
+      directory = "plugin"
+      requires = "rocq-core.plugins.ltac rocq-core.vernac rocq-elpi.plugin"
+      ...
+    )
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.tc.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  ...
+)
+```
+
+# Drawbacks
+
+None, except the eventual need to port all projects to use a build system that
+respects the new installation scheme. However, as a first step, it is possible
```

Comment:
Alternatively, if the build systems are ported, maybe there will be no porting needed on project side.

Reply:
Actually, I think that porting will be required regardless: the build systems will need to be provided with different information (basically a list of package names the project depends on). For `dune`, this information is already required in some form, but that might need to change. In particular, theories that are dependencies currently need to be listed as directory paths, and these would need to become package names. (Furthermore, the `(plugins ...)` and `(theories ...)` fields of the `coq.theory` stanza could be merged in a single `(deps ...)` field.)

In any case, care needs to be taken to ensure that a smooth transition is possible.

Reply:
The best transition is one that allows using either the old or the new build method for a transition period.  That may be difficult.  Requiring immediate but very simple changes to every interesting Rocq package in opam could be a bit of a rocky transition.